float32 xb
float32 yb



